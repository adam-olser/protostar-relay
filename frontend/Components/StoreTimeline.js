import React, { useState, useEffect } from "react";
import Slider from 'react-input-slider';


function StoreTimeline() {
    const [state, setState] = useState({ x: 0.3 });

    return (
        <div>
            <div>
                <code>
                    <p>insert code here...</p>
                    {/* <pre>
                        type Query {
                            getMessage(id: ID!): Message
                        }
                    </pre> */}
                </code>
            </div>
            <div>
                <p>addUser 1</p>
                <p>addUser 2</p>
            </div>
            <div>
                {/* Need to research what slider to use */}
                <Slider
                    axis="x"
                    xstep={0.1}
                    xmin={0}
                    xmax={1}
                    x={state.x}
                    onChange={({ x }) => setState({ x: parseFloat(x.toFixed(2)) })}
                />
            </div>
        </div>
    )
}

export default StoreTimeline;