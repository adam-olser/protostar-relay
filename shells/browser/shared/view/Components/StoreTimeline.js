import React, { useState, useEffect } from "react";
import Slider from 'react-input-slider';


function StoreTimeline() {
    const [state, setState] = useState({ x: 0.3 });

    return (
        <div>
            <div className="display-box">
                {/* {
                    type Query {
                        getMessage(id: ID!): Message
                    }
                } */}
                code goes here....
            </div>
            <div>
                <p>addUser 1</p>
                <p>addUser 2</p>
            </div>
            <div>
                <Slider
                    axis="x"
                    xstep={0.1}
                    xmin={0}
                    xmax={1}
                    x={state.x}
                    onChange={({ x }) => setState({ x: parseFloat(x.toFixed(2)) })}
                />
                <h2 className="slider-textcolor">Store Timeline</h2>
            </div>
        </div>
    )
}

export default StoreTimeline;