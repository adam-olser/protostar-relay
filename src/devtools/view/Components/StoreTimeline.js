import React, { useState, useEffect } from "react";
import Slider from 'react-input-slider';


const StoreTimeline = ({store}) => {
  const [state, setState] = useState({ x: 0.3 });
  const [timeline, setTimeline] = useState([]);
  const [timelineLabel, setTimelineLabel] = useState('');

  const handleClick = (e) => {
    e.preventDefault();
    const timelineInsert = {};
    const timeStamp = Date.now();
    const newStore = store;
    timelineInsert.label = timelineLabel;
    timelineInsert.date = timeStamp;
    timelineInsert.store = newStore;
    setTimeline([...timeline, timelineInsert]);
  }

    return (
        <div>
            <div className="display-box">
                  
                {/* {
                    type Query {
                        getMessage(id: ID!): Message
                    }
                } */}
                code goes here....
                <input type="text" value={timelineLabel} onChange={(e) => setTimelineLabel(e.target.value)} ></input>
                <a className="button mx-1" onClick={(e) => handleClick(e)}>Timeline snapshot</a>
              <div>{timeline.length}</div>
              <div>{timelineLabel}</div>
            </div>
            <div>
                <p>addUser 1</p>
                <p>addUser 2</p>
            </div>
            <h2 className="slider-textcolor">Store Timeline</h2>
        </div>
    )
}

export default StoreTimeline;