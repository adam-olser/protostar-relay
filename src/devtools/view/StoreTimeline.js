import React, { useState, useEffect, useContext, useCallback } from "react";
import InputRange from "react-input-range";
import { BridgeContext, StoreContext } from "../context";
import StoreDisplayer from "./StoreDisplayer";
// import { detailedDiff } from 'deep-object-diff';

const StoreTimeline = ({ currentEnvID }) => {
  const store = useContext(StoreContext);
  const bridge = useContext(BridgeContext);
  const [, forceUpdate] = useState({});
  const [snapshotIndex, setSnapshotIndex] = useState(0);
  const [timelineLabel, setTimelineLabel] = useState("");
  const [liveStore, setLiveStore] = useState({});
  const [timeline, setTimeline] = useState({
    [currentEnvID]: [
      {
        label: "at startup",
        date: Date.now(),
        storage: liveStore,
      },
    ],
  });

  const handleClick = (e) => {
    e.preventDefault();
    const timelineInsert = {};
    const timeStamp = Date.now();
    timelineInsert.label = timelineLabel;
    timelineInsert.date = timeStamp;
    timelineInsert.storage = liveStore;
    const newTimeline = timeline[currentEnvID].concat([timelineInsert]);
    setTimeline({ ...timeline, [currentEnvID]: newTimeline });
    setTimelineLabel("");
    setSnapshotIndex(newTimeline.length);
  };

  const updateStoreHelper = (storeObj) => {
    setLiveStore(storeObj);
  };
  // testing sans set timeout... maybe put back
  useEffect(() => {
    const refreshLiveStore = () => {
      console.log("mutation triggered refr");
      bridge.send("refreshStore", currentEnvID);
    };
    const refreshEvents = () => {
      console.log("storeinspector refreshing due to storeDataReceived flag");
      const allRecords = store.getRecords(currentEnvID);
      updateStoreHelper(allRecords);
      console.log("REFRESH INVOKED! currentEnvID:", currentEnvID);
      // forceUpdate({});
    };

    store.addListener("storeDataReceived", refreshEvents);
    store.addListener("allEventsReceived", refreshEvents);
    store.addListener("mutationComplete", refreshLiveStore);

    return () => {
      store.removeListener("mutationComplete", refreshLiveStore);
      store.removeListener("storeDataReceived", refreshEvents);
      store.removeListener("allEventsReceived", refreshEvents);
    };
  }, [store]);

  useEffect(() => {
    const allRecords = store.getRecords(currentEnvID);
    setLiveStore(allRecords);

    if (!timeline[currentEnvID]) {
      const newTimeline = {
        ...timeline,
        [currentEnvID]: [
          {
            label: "current",
            date: Date.now(),
            storage: allRecords,
          },
        ],
      };
      setTimeline(newTimeline);
      setSnapshotIndex(1);
    } else {
      setSnapshotIndex(timeline[currentEnvID].length);
    }
  }, [currentEnvID]);

  console.log("Rendering StoreTimeline");
  console.log("livestore", liveStore);

  return (
    <React.Fragment>
      <div className="column">
        <div className="display-box">
          <div className="snapshot-wrapper is-flex">
            <input
              type="text"
              className="input is-small snapshot-btn is-primary"
              value={timelineLabel}
              onChange={(e) => setTimelineLabel(e.target.value)}
              placeholder="take a store snapshot"
            ></input>
            <button
              className="button is-small is-link"
              onClick={(e) => handleClick(e)}
            >
              Snapshot
            </button>
          </div>
        </div>
        <div className="snapshots">
          <h2 className="slider-textcolor">Store Timeline</h2>
          <InputRange
            maxValue={
              timeline[currentEnvID] ? timeline[currentEnvID].length : 0
            }
            minValue={0}
            value={snapshotIndex}
            onChange={(value) => setSnapshotIndex(value)}
          />
          <div className="snapshot-nav">
            <button
              class="button is-small"
              onClick={() => {
                if (snapshotIndex !== 0) setSnapshotIndex(snapshotIndex - 1);
              }}
            >
              Backward
            </button>
            <button
              class="button is-small"
              onClick={() => setSnapshotIndex(timeline[currentEnvID].length)}
            >
              Current
            </button>
            <button
              class="button is-small"
              onClick={() => {
                if (snapshotIndex !== timeline[currentEnvID].length)
                  setSnapshotIndex(snapshotIndex + 1);
              }}
            >
              Forward
            </button>
          </div>
          <div className="snapshot-info"></div>
        </div>
      </div>
      <StoreDisplayer
        store={
          !timeline[currentEnvID] ||
          !timeline[currentEnvID][snapshotIndex] ||
          snapshotIndex === timeline[currentEnvID].length
            ? liveStore
            : timeline[currentEnvID][snapshotIndex].storage
        }
      />
    </React.Fragment>
  );
};

export default StoreTimeline;
