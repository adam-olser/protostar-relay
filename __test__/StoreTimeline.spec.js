import React from 'react';
import { configure, shallow, render } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import renderer from 'react-test-renderer';
import StoreTimeline from '../src/devtools/view/StoreTimeline';

configure({ adapter: new Adapter() });

describe('StoreTimeline', () => {
  let wrapper;
  const props = {
    store: {
      '1': {
        '__id': '1',
        '__typename': 'User',
        'name': 'Marc'
      },
      '2': {
        '__id': '2',
        '__typename': 'User',
        'name': 'Aryeh'
      },
      '3': {
        '__id': '3',
        '__typename': 'User',
        'name': 'Liz'
      },
      '4': {
        '__id': '4',
        '__typename': 'User',
        'name': 'Qwen'
      },
      '5': {
        '__id': '5',
        '__typename': 'Post',
        'text': 'Hi'
      },

    }
  }

  beforeAll(() => {
    wrapper = shallow(<StoreTimeline {...props} />);
  });

  it("Has a StoreDisplayer component", () => {

  })

  it("Passes the store as a prop to StoreDisplayer", () => {

  })

  it("Passes the store based on the currentEnvID", () => {

  })


  describe("Snapshots", () => {

    it("Takes a snapshot at startup", () => {

    })

    it("Has a snapshot button that takes and saves a snapshot", () => {

    })

    it("Defaults to displaying the latest store value when a snapshot is taken", () => {

    })

    it("Remembers snapshots when switching between environments", () => {

    })

    it("Has a snapshot text input", () => {

    })

    it("Has a previous buttons to move to the previous snapshot", () => {

    })

    it("Has a next button to move to the previous snapshot", () => {

    })

    it("Has a current button that shows the current store value", () => {

    })

    it("Has a slider that updates when a new snapshot is taken and when switching between environments", () => {

    })


  })

  it('Renders correctly', () => {
    const tree = renderer.create(<StoreTimeline {...props} />).toJSON();
    expect(tree).toMatchSnapshot();
  })

});