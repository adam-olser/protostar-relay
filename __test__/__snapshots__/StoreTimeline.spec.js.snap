// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StoreTimeline Renders correctly 1`] = `
Array [
  <div
    className="column"
  >
    <div
      className="display-box"
    >
      <div
        className="snapshot-wrapper is-flex"
      >
        <input
          className="input is-small snapshot-btn is-primary"
          onChange={[Function]}
          placeholder="take a store snapshot"
          type="text"
          value=""
        />
        <button
          className="button is-small is-link"
          onClick={[Function]}
        >
          Snapshot
        </button>
      </div>
    </div>
    <div
      className="snapshots"
    >
      <h2
        className="slider-textcolor"
      >
        Store Timeline
      </h2>
      <div
        aria-disabled={false}
        className="input-range"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
      >
        <span
          className="input-range__label input-range__label--min"
        >
          <span
            className="input-range__label-container"
          >
            0
          </span>
        </span>
        <div
          className="input-range__track input-range__track--background"
          onMouseDown={[Function]}
          onTouchStart={[Function]}
        >
          <div
            className="input-range__track input-range__track--active"
            style={
              Object {
                "left": "0%",
                "width": "0%",
              }
            }
          />
          <span
            className="input-range__slider-container"
            style={
              Object {
                "left": "0%",
                "position": "absolute",
              }
            }
          >
            <span
              className="input-range__label input-range__label--value"
            >
              <span
                className="input-range__label-container"
              >
                0
              </span>
            </span>
            <div
              aria-valuemax={1}
              aria-valuemin={0}
              aria-valuenow={0}
              className="input-range__slider"
              draggable="false"
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              onTouchStart={[Function]}
              role="slider"
              tabIndex="0"
            />
          </span>
        </div>
        <span
          className="input-range__label input-range__label--max"
        >
          <span
            className="input-range__label-container"
          >
            1
          </span>
        </span>
      </div>
      <div
        className="snapshot-nav"
      >
        <button
          class="button is-small"
          onClick={[Function]}
        >
          Backward
        </button>
        <button
          class="button is-small"
          onClick={[Function]}
        >
          Current
        </button>
        <button
          class="button is-small"
          onClick={[Function]}
        >
          Forward
        </button>
      </div>
      <div
        className="snapshot-info"
      />
    </div>
  </div>,
  <div
    className="column"
  >
    <p
      class="control has-icons-left"
    >
      <input
        className="input is-small is-primary"
        onChange={[Function]}
        placeholder="Search"
        type="text"
      />
      <button
        className="button is-small is-link"
        onClick={[Function]}
      >
        Reset
      </button>
      <span
        class="icon is-left"
      >
        <i
          class="fas fa-search"
        />
      </span>
    </p>
    <aside
      className="menu"
    >
      <p
        className="menu-label"
      >
        Record List
      </p>
      <ul
        className="menu-list"
      />
    </aside>
  </div>,
  <div
    className="column"
  >
    <div
      className="display-box"
    >
      <div
        className="records"
      />
    </div>
  </div>,
]
`;
